File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def bfs(maps, x,y):
    q = deque([(x,y)]) #시작 위치
    
    while q:
        tmpX, tmpY = q.popleft()
        for i in range(4):
            nx = tmpX + dx[i]
            ny = tmpY + dy[i]
            
            if 0<=nx<n and 0<=ny<m and maps[nx][ny] == 1: #갈 수 있음
                q.append((nx, ny))
                dist[nx][ny] = dist[tmpX][tmpY]+1 #거리 기록
                maps[nx][ny] = 0 #방문 기록
        

def solution(maps):
    answer = 0
    global dx,dy,n,m, dist
    
    dx = [0, 0, 1,-1] #동서남북
    dy = [1,-1, 0, 0]
    
    n = len(maps) #가로
    m = len(maps[0]) #세로
    
    dist = [[0]*m for _ in range(n)]
    
    bfs(maps, 0,0)
    print(dist)
    
    if dist[n-1][m-1] == 0:
        return -1
    else:
        return dist[n-1][m-1]+1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function bfs(maps) {
    const n = maps.length; // 가로 길이
    const m = maps[0].length; // 세로 길이

    const dx = [0,1,0,-1]
    const dy = [1,0,-1,0];
    const dist = Array.from({ length: n }, () => Array(m).fill(0)); // 거리 기록
    const queue = [[0, 0]]; 
    dist[0][0] = 1; 

    while (queue.length > 0) {
        const [x, y] = queue.shift(); 

        for (let i = 0; i < 4; i++) {
            const nx = x + dx[i];
            const ny = y + dy[i];

            if (nx >= 0 && nx < n && ny >= 0 && ny < m && maps[nx][ny] === 1) {
                queue.push([nx, ny]); 
                dist[nx][ny] = dist[x][y] + 1; 
                maps[nx][ny] = 0; 
            }
        }
    }

    return dist;
}

function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;

    const dist = bfs(maps);

    if (dist[n-1][m-1] === 0) return -1;
    

    return dist[n-1][m-1];
}
